src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionProvider.cs

```diff
@@ -42,32 +46,119 @@ protected override async Task AddCompletionItemsAsync(CompletionContext completi
+                    // so just ignore the Bar<int> here.
+                    var typeParameter = target.GetTypeParameters();
+                    if (typeParameter.IsEmpty)
+                    {
+                        var namespaceOfTarget = target.ContainingNamespace.ToDisplayString(SymbolDisplayFormats.NameFormat);
+                        var typeNameOfTarget = target.Name;
+                        dictionary.Add(namespaceOfTarget, typeNameOfTarget);
+                    }
+                }
+            }
+
+            return dictionary;
+        }
+
+        private static void AddItems(
+            ImmutableArray<CompletionItem> items,
+            CompletionContext completionContext,
+            HashSet<string> namespacesInScope,
+            MultiDictionary<string, string> aliasTargetNamespaceToTypeNameMap,
+            TelemetryCounter counter)
         {
             counter.ReferenceCount++;
             foreach (var item in items)
             {
-                var containingNamespace = ImportCompletionItem.GetContainingNamespace(item);
-                if (!namespacesInScope.Contains(containingNamespace))
+                if (ShouldAddItem(item, namespacesInScope, aliasTargetNamespaceToTypeNameMap))
                 {
                     // We can return cached item directly, item's span will be fixed by completion service.
                     // On the other hand, because of this (i.e. mutating the  span of cached item for each run),
-                    // the provider can not be used as a service by components that might be run in parallel 
+                    // the provider can not be used as a service by components that might be run in parallel
                     // with completion, which would be a race.
                     completionContext.AddItem(item);
                     counter.ItemsCount++;
                 }
             }
+
+            static bool ShouldAddItem(
+                CompletionItem item,
+                HashSet<string> namespacesInScope,
+                MultiDictionary<string, string> aliasTargetNamespaceToTypeNameMap)
+            {
+                var containingNamespace = ImportCompletionItem.GetContainingNamespace(item);
+                // 1. if the namespace of the item is in scoop. Don't add the item
+                if (namespacesInScope.Contains(containingNamespace))
+                {
+                    return false;
+                }
+
+                // 2. If the item might be an alias target. First check if the target alias map has any value then
+                // check if the type name is in the dictionary.
+                // It is done in this way to avoid calling ImportCompletionItem.GetTypeName for all the CompletionItems
+                if (!aliasTargetNamespaceToTypeNameMap.IsEmpty
+                    && aliasTargetNamespaceToTypeNameMap[containingNamespace].Contains(ImportCompletionItem.GetTypeName(item)))
```

Comment:
nit: I guess we really don't care if it's a generic type, since no generic target will be in the dictionary.

Reply:
why do you need the aliasTargetNamespaceToTypeNameMap.IsEmpty check?

Reply:
Because it cheaper than indexing into the dictionary. We are dealing with large amount of item and it's very rare there's using alias in the document